db-engine/src/main/java/com/splicemachine/db/iapi/types/SQLDecfloat.java

```diff
@@ -0,0 +1,995 @@
+     * decimal point, or zero for a whole number. This does not match the
+     * return from BigDecimal.scale() since in J2SE 5.0 onwards that can return
+     * negative scales.
+     */
+    public int getDecimalValuePrecision()
+    {
+        if (isNull())
+            return 0;
+
+        BigDecimal localValue = getBigDecimal();
+
+        return SQLDecfloat.getWholeDigits(localValue) + getDecimalValueScale();
+    }
+
+    /**
+     * Get a BigDecimal representing the value of a DataValueDescriptor
+     * @param value Non-null value to be converted
+     * @return BigDecimal value
+     * @throws StandardException Invalid conversion or out of range.
+     */
+    public static BigDecimal getBigDecimal(DataValueDescriptor value) throws StandardException
+    {
+        if (SanityManager.DEBUG)
+        {
+            if (value.isNull())
+                SanityManager.THROWASSERT("NULL value passed to SQLDecimal.getBigDecimal");
+        }
+
+        switch (value.typeToBigDecimal())
+        {
+        case com.splicemachine.db.iapi.reference.Types.DECFLOAT:
+        case Types.DECIMAL:
+            return ((BigDecimal) value.getObject()).round(MathContext.DECIMAL128);
+        case Types.CHAR:
+            try {
+                return new BigDecimal(value.getString().trim(), MathContext.DECIMAL128);
+            } catch (NumberFormatException nfe) {
+                throw StandardException.newException(SQLState.LANG_FORMAT_EXCEPTION, "java.math.BigDecimal");
+            }
+        case Types.BIGINT:
+            return BigDecimal.valueOf(value.getLong());
+        default:
+            if (SanityManager.DEBUG)
+                SanityManager.THROWASSERT("invalid return from " + value.getClass() + ".typeToBigDecimal() " + value.typeToBigDecimal());
+            return null;
+        }
+    }
+
+    /**
+     * Calculate the number of digits to the left of the decimal point
+     * of the passed in value.
+     * @param decimalValue Value to get whole digits from, never null.
+     * @return number of whole digits.
+     */
+    private static int getWholeDigits(BigDecimal decimalValue) {
+        /**
+         * if ONE > abs(value) then the number of whole digits is 0
+         */
+        decimalValue = decimalValue.abs();
+        if (ONE.compareTo(decimalValue) > 0)
+            return 0;
+        return decimalValue.precision() - decimalValue.scale();
+    }
+
+    public Format getFormat() {
+        return Format.DECFLOAT;
+    }
+
+    @Override
+    public void read(Row row, int ordinal) throws StandardException {
+        if (row.isNullAt(ordinal))
+            setToNull();
+        else {
+            isNull = false;
+            value = row.getDecimal(ordinal).round(MathContext.DECIMAL128);
+        }
+    }
+
+    @Override
+    public StructField getStructField(String columnName) {
+        return DataTypes.createStructField(columnName, DataTypes.BinaryType, true);
```

Comment:
It may be not right to use DataTypes.BinaryType here. For native spark execution path, the use of defloat type will fail for the following case:
```
create table t1 (a1 decfloat, b1 int);
create table t2 (a2 decfloat, b2 int);
insert into t1 values (3,3);
insert into t2 values (3,3);

select * from t1, t2 --splice-properties useSpark=true, joinStrategy=broadcast
> where a1=a2;
ERROR SE001: Splice Engine exception: unexpected exception
ERROR XJ001: Java exception: 'java.lang.RuntimeException: java.math.BigDecimal is not a valid external type for schema of binary: java.io.IOException'.
ERROR XJ001: Java exception: 'java.math.BigDecimal is not a valid external type for schema of binary: java.lang.RuntimeException'.
```

Reply:
Good point,
SQLDecfloat.getSparkObject / setSparkObject now work with Strings. 
Until we define a proper native spark sql type for decfloat, ordering will not be possible